### ISSUE TITLE
`bash -c fish` hanging indefinitely

### ISSUE THREAD

**Issue by mqudsi** [Microsoft/WSL#1653]
Attempting to start my upgraded installation of fish shell via bash.exe (`bash -c fish`) causes bash.exe to hang indefinitely. I have attached a full dump of bash.exe in this state: [bash.zip](https://github.com/Microsoft/BashOnWindows/files/745175/bash.zip)

This is on Windows 10 build 15014.

Launching bash then executing `fish` works just fine. Launching any other shell via `bash -c ...` (e.g. `sh`) also works just fine.

This did not use to happen until I updated my fish installation.

    mqudsi@ZBook /m/c/U/Mahmoud> fish --version
    fish, version 2.4.0-472-gf48291b

**Comment by benhillis** [comment]
I can't reproduce this on [Version 10.0.15026].  Are there any other steps specific to your configuration?

This is what I did:
1. sudo apt install fish
2. bash.exe -c fish
```
C:\Users\benhill>bash.exe -c fish
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
benhill@BENHILL-X1 /m/c/U/benhill>
```

**Comment by mqudsi** [comment]
Yes, fish worked fine when installed that way for me too. However, when I manually compiled it from git to run the latest version, I ran into the problem described above.

Since I think bash on Windows is ABI stable, I've attached a copy of my own fish binary here. Otherwise, you can build it yourself from https://github.com/fish-shell/fish-shell

[fish.gz](https://github.com/Microsoft/BashOnWindows/files/745622/fish.gz)

**Comment by therealkenc** [comment]
Here's the [gist](https://gist.github.com/therealkenc/3776406f80318db15db09bb3e8f30cd6) for version `2.4.0-477-g44cfe3e`, tail end below. It only hangs when run from `bash -c`. Runs fine if launched from bash inside WSL. Code in question seems to be [here](https://github.com/fish-shell/fish-shell/blob/5b6cc5af6d5531f6fa3506da4e4b11cc6b60fca5/src/reader.cpp#L1627-L1632). I bisected a little, but just to make life difficult that `tcsetpgrp()` appears to be in 2.2.0 (which works fine) as well.

Note this exited below because of ctrl-c, otherwise it just hangs forever.
```
ioctl(1, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(1, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(1, TIOCGWINSZ, {ws_row=47, ws_col=120, ws_xpixel=0, ws_ypixel=0}) = 0
getpgrp()                               = 225
ioctl(0, TIOCGPGRP, [225])              = 0
getpgrp()                               = 225
setpgid(228, 228)                       = 0
ioctl(0, TIOCSPGRP, [228])              = -1 EINTR (Interrupted system call)
--- SIGTTOU {si_signo=SIGTTOU, si_code=SI_KERNEL} ---
strace: Process 228 detached
<E> fish: Couldn't grab control of terminal
tcsetpgrp: Interrupted system call
```

**Comment by vb4life** [comment]
Might not be related, but after I upgraded to `fish, version 2.6.0` it started hanging when launched. The difference is that I added fish to my `.bashrc`. If i started it manually, everything was fine.

I managed to solve it by just creating an empty fish.config file:
`# touch ~/.config/fish/config.fish`

This might be a completely different issue, but was the only thing a quick google search could find :)

**Comment by mqudsi** [comment]
@benhillis any chance of an update on this? Do you think it's related to the other `fish` issues?

**Comment by mqudsi** [comment]
Just an update: the latest `fish` compiled from git does not exhibit this problem; but not because the issue went away. Bisecting the changes between `g44cfe3e` and `b88cacf03eb5717cf2ecaaa6ab4632bc7f7632d5` and comparing that to the previous bisect logs for when the problem was introduced (aside from being a nightmare) should be enlightening.

**Comment by therealkenc** [comment]
Curious. Nothing obvious changed in fish with respect to the [relevant code](https://github.com/fish-shell/fish-shell/blame/aec097319646bae50dfa70ab31b102f188cdb92b/src/reader.cpp#L1650) recently.

Interestingly `fish` does have a test for WSL [here](https://github.com/fish-shell/fish-shell/blob/cf808674bf2be527c92895af0a681c597455b6be/src/env.cpp#L741). But it's unrelated AFAICT.

**Comment by mqudsi** [comment]
I was mistaken previously. It seems to be unrelated to the fish code.

I experienced the hang again after making some changes to fish, and wanted to `git bisect` to see what caused it, but I found that the commit I previously reported above was _also_ broken for me now.

I haven't installed any Windows updates, I simply recompiled `fish` and now the problem is back. I thought maybe it was a compiler issue (I compiled with clang++ 4.0 whereas previously was with g++ vSomething) but recompiling with g++7 kept the problem.

**Comment by mqudsi** [comment]
OK, so it's not something in the fish source code. I updated to the latest fish code and rebuilt and installed only for it to begin hanging once more. I decided to `git bisect` the problem, except reverting to the commit I noted in my previous build as working did _not_ work this time around.

I'm having a hard time figuring this one out since running with full debug logging doesn't reveal anything interesting. The diff between the successful start and the failed start is also not very interesting.

Trying to debug with `gdb` causes the hung `fish` process to crash on attach, before `gdb` gets a chance to hook in and results in no stack trace. Launching fish under gdb hangs both gdb and fish when launched and neither responds to `SIGINT` .

I thought it might have something to do with the compiler; last time around I compiled with g++ but I'm not sure which version, while this time was with clang++ 4.0; so I tried to rebuild with g++ 7, but that wasn't it.

**Comment by mqudsi** [comment]
I tracked it down to the following code:

```cpp
    debug(0, L"calling tcsetpgrp for fd %d for pgid %d\n", STDIN_FILENO, shell_pgid);
    if (tcsetpgrp(STDIN_FILENO, shell_pgid) == -1) {
        debug(0, L"tcsetpgrp failed\n");
        if (errno == ENOTTY) redirect_tty_output();
        debug(0, _(L"Couldn't grab control of terminal"));
        wperror(L"tcsetpgrp");
        exit_without_destructors(1);
    }
    debug(0, L"tcsetpgrp ok\n");
```

This hangs after printing "calling tcsetpgrp ..."

**Comment by lepinekong** [comment]
I confirm that it freezes.

**Comment by mathewhany** [comment]
I don't know if this could help anyone, but I managed to workaround this problem, I'm using Hyper as a terminal, and what I did is:

1. I followed the installation guide to install fish.
```
sudo apt-add-repository ppa:fish-shell/release-2
sudo apt-get update
sudo apt-get install fish
```
2. Then, I installed Oh My Fish (However, I don't think installing it made any difference to the problem).
```
curl -L https://get.oh-my.fish | fish
```
3. I edited `.bashrc`, I added at the end:
```
if [ -t 1 ]; then
  exec fish
fi
```
4. In `.hyper.js`, I edited it like this:
```
shell: 'bash.exe',
shellArgs: ['--login'],
```
